index.ts

```diff
@@ -51,46 +83,51 @@ export const setupPlugin: S3Plugin['setupPlugin'] = ({ global, config }) => {
-            const date = new Date().toISOString()
-            const [day, time] = date.split('T')
-            const dayTime = `${day.split('-').join('')}-${time.split(':').join('')}`
-            const suffix = randomBytes(8).toString('hex')
-
-            const params = {
-                Bucket: config.s3BucketName,
-                Key: `${config.prefix || ''}${day}/${dayTime}-${suffix}.jsonl`,
-                Body: Buffer.from(batch.map((event) => JSON.stringify(event)).join('\n'), 'utf8'),
-            }
-
-            if (config.compression === 'gzip') {
-                params.Key = `${params.Key}.gz`
-                params.Body = gzipSync(params.Body)
-            }
-
-            if (config.compression === 'brotli') {
-                params.Key = `${params.Key}.br`
-                params.Body = brotliCompressSync(params.Body)
-            }
-            global.s3.upload(params, (err: Error, data: ManagedUpload.SendData) => {
-                if (err) {
-                    console.error(`Error uploading to S3: ${err.message}`)
-                    throw err
-                }
-                console.log(`Uploaded ${batch.length} event${batch.length === 1 ? '' : 's'} to ${data.Location}`)
-            })
+        timeoutSeconds: uploadMinutes * 60, 
+        onFlush: async (batch) => {
+            await jobs.uploadBatchToS3({ batch, batchId: Math.floor(Math.random() * 1000000), retriesPerformedSoFar: 0 }).runNow()
         },
     })
+
     global.eventsToIgnore = new Set(
         config.eventsToIgnore ? config.eventsToIgnore.split(',').map((event) => event.trim()) : null
     )
 }
 
-export const processEvent: S3Plugin['processEvent'] = (event, { global }) => {
+export const onEvent = (event: PluginEvent, { global }: S3Meta) => {
     if (!global.eventsToIgnore.has(event.event)) {
         global.buffer.add(event)
     }
-    return event
+}
+
+export const sendBatchToS3 = (batch: PluginEvent[], { global, config }: S3Meta) => {
+    const date = new Date().toISOString()
+    const [day, time] = date.split('T')
+    const dayTime = `${day.split('-').join('')}-${time.split(':').join('')}`
+    const suffix = randomBytes(8).toString('hex')
+
+    const params = {
+        Bucket: config.s3BucketName,
+        Key: `${config.prefix || ''}${day}/${dayTime}-${suffix}.jsonl`,
+        Body: Buffer.from(batch.map((event) => JSON.stringify(event)).join('\n'), 'utf8'),
+    }
+
+    if (config.compression === 'gzip') {
+        params.Key = `${params.Key}.gz`
+        params.Body = gzipSync(params.Body)
+    }
+
+    if (config.compression === 'brotli') {
+        params.Key = `${params.Key}.br`
+        params.Body = brotliCompressSync(params.Body)
+    }
+
+    console.log(`Flushing ${batch.length} events!`)
+    global.s3.upload(params, (err: Error, data: ManagedUpload.SendData) => {
+        if (err) {
+            console.error(`Error uploading to S3: ${err.message}`)
+            throw new UploadError()
```

Comment:
Is this error actually captured, or does the callback run async in the background and it'll be swollen?

Reply:
hmm that's a good point - this did work locally but I should check the implementation and await this if needed